PersistentSessionsWorker: retry with 0 backoff ms.

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

We released keycloak 26 and functional and performance wise everything is good.

However around once a day a duplicate insert session happens , which results in 20.000 error messages in a 10 second time frame.

We have reduced this number by overwriting PersistentSessionsWorker, with a back-off time of 100ms instead of "0" (during 5 instead of 10) which is in the code:

   Retry.executeWithBackoff((iteration) -> {
   ....
   }, Duration.of(10L, ChronoUnit.SECONDS), 0);
   
   
   
   The duplicate itself concerns 1 specific session id (1 of the 10.000 ?) around once a day, and doesn't seem to be a functional big issue: 
   
   HHH100501: Exception executing batch [java.sql.BatchUpdateException: Duplicate entry 'b9f91c53-ec37-4aaf-b54d-12f793c58d09-1' for key 'OFFLINE_USER_SESSION.PRIMARY'], SQL: insert into OFFLINE_USER_SESSION (BROKER_SESSION_ID,CREATED_ON,DATA,LAST_SESSION_REFRESH,REALM_ID,USER_ID,VERSION,OFFLINE_FLAG,USER_SESSION_ID) values (?,?,?,?,?,?,?,?,?)
   
   
   (repeated 20.000 times of same entry id.)
   
   
   
   

### Version

26.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

when retry a insert session statement or others, use a backoff time > 0ms to avoid a bulk of same error messages.


### Actual behavior

To many error messages because of back-off delay 0ms.


### How to Reproduce?

generate a sql exception (like duplicate insert).



### Anything else?

_No response_

The issue we have seems to be similar.
We run Keycloak 26.0.7 on kubernetes with cloud-native postgres.

These are the ENV vars:
KC_CACHE_STACK: "kubernetes"
CACHE_OWNERS_COUNT: "2"
CACHE_OWNERS_AUTH_SESSIONS_COUNT: "2"

We have 3 instances of cloudnative-pg/postgresql:15.10

In the logging we see a log of these exceptions (224 between 2025-01-06 20:36:57,524 and 2025-01-08 09:19:25,147):
```
2025-01-06 20:37:26,789 ERROR [org.hibernate.orm.jdbc.batch] (org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker) HHH100501: Exception executing batch [org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?], SQL: update OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?
2025-01-06 20:37:26,789 WARN  [com.arjuna.ats.arjuna] (org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker) ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffff0ae960fb:b131:6757092c:4a791f, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@3f33a1b4 >: jakarta.persistence.OptimisticLockException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?
	at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:221)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:95)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:167)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:173)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1433)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:491)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2354)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1978)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:336)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
	at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:348)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:66)
	at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:135)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1290)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
	at io.quarkus.narayana.jta.runtime.NotifyingTransactionManager.commit(NotifyingTransactionManager.java:70)
	at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:100)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
	at org.keycloak.services.DefaultKeycloakSession.closeTransactionManager(DefaultKeycloakSession.java:396)
	at org.keycloak.services.DefaultKeycloakSession.close(DefaultKeycloakSession.java:361)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:439)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:312)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:302)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.lambda$process$3(PersistentSessionsWorker.java:89)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:102)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:93)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.process(PersistentSessionsWorker.java:86)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.run(PersistentSessionsWorker.java:69)
Caused by: org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?
	at org.hibernate.jdbc.Expectations.checkBatched(Expectations.java:80)
	at org.hibernate.jdbc.Expectation$RowCount.verifyOutcome(Expectation.java:174)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.checkRowCounts(BatchImpl.java:321)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$2(BatchImpl.java:285)
	at org.hibernate.engine.jdbc.mutation.internal.PreparedStatementGroupSingleTable.forEachStatement(PreparedStatementGroupSingleTable.java:67)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.performExecution(BatchImpl.java:264)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.execute(BatchImpl.java:242)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.executeBatch(JdbcCoordinatorImpl.java:188)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:674)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:511)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:414)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:41)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1429)
	... 26 more
```

The behaviour in our metrics of the db wal is the following:
![image](https://github.com/user-attachments/assets/87b7053b-7e90-44ed-8c1f-b725f7300cef)

orange (db-wal-1) and yellow (db-wal-2) are trying to sync with db-wal-3 (blue), but can't, since there is some lock.

Probably some issue is introduced when keeping the offline sessions in DB instead of in cache.

--- a/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java
+++ b/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java
@@ -24,6 +24,7 @@ import io.opentelemetry.context.Context;
 import org.jboss.logging.Logger;
 import org.keycloak.common.util.Retry;
 import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.ModelDuplicateException;
 import org.keycloak.models.utils.KeycloakModelUtils;
 import org.keycloak.tracing.TracingProvider;
 
@@ -119,8 +120,20 @@ public class PersistentSessionsWorker {
                                                         change::perform);
                                                 change.complete();
                                                 performedChanges.add(change);
+                                            } catch (ModelDuplicateException ex) {
+                                                // duplicate exceptions are unlikely to succeed on a retry,
+                                                tracing.error(ex);
+                                                change.fail(ex);
+                                                performedChanges.add(change);
                                             } catch (Throwable ex) {
-                                                throwables.add(ex);
+                                                if (iteration > 20) {
+                                                    // never retry more than 20 times
+                                                    tracing.error(ex);
+                                                    change.fail(ex);
+                                                    performedChanges.add(change);
+                                                } else {
+                                                    throwables.add(ex);
+                                                }
                                             }
                                         });
                                         batch.removeAll(performedChanges);

